File: control-plane/src/server/mcp-tools/queryFleet.ts
import { defineTool } from "@github/copilot-sdk";
import { z } from "zod";
import type { StateStore, WorkflowFilters } from "@server/services/stateStore";
import type { PhaseName } from "@shared/types";

export function queryFleetTool(store: StateStore) {
  return defineTool("query-fleet", {
    description: "Aggregated fleet state — returns workflow counts by phase/status and recent open exceptions.",
    parameters: z.object({
      phase: z.string().optional().describe("Filter by current phase name"),
      agency: z.string().optional().describe("Filter by agency name"),
      hasException: z.boolean().optional().describe("Filter to workflows that have an active exception"),
    }),
    skipPermission: true,
    handler: async ({ phase, agency, hasException }) => {
      const filters: WorkflowFilters = {
        phase: phase as PhaseName | undefined,
        agency,
        hasException,
      };
      const workflows = store.listWorkflows(filters);
      const allExceptions = store.listExceptions();

      const byPhase: Record<string, number> = {};
      const byStatus: Record<string, number> = {};
      for (const w of workflows) {
        byPhase[w.currentPhase] = (byPhase[w.currentPhase] ?? 0) + 1;
        byStatus[w.status] = (byStatus[w.status] ?? 0) + 1;
      }

      const recentExceptions = allExceptions
        .sort((a, b) => b.createdAt - a.createdAt)
        .slice(0, 5)
        .map(e => ({
          id: e.id,
          workflowId: e.workflowId,
          severity: e.severity,
          category: e.category,
          summary: e.summary,
          confidence: e.confidence,
        }));

      return {
        total: workflows.length,
        byPhase,
        byStatus,
        openExceptionCount: allExceptions.length,
        recentExceptions,
      };
    },
  });
}

File: control-plane-py/src/server/mcp_tools/query_fleet.py
# src/server/mcp_tools/query_fleet.py
from __future__ import annotations
import json
from pydantic import BaseModel, Field
from copilot.tools import define_tool, ToolInvocation, ToolResult
from src.server.services.state_store import StateStore


class QueryFleetParams(BaseModel):
    phase: str | None = Field(default=None, description="Filter to a specific workflow phase")
    agency: str | None = Field(default=None, description="Filter to a specific agency")
    has_exception: bool | None = Field(default=None, description="Filter to workflows with active exceptions")


def make_query_fleet_tool(store: StateStore):
    @define_tool(description="Aggregated state of the workflow fleet.", skip_permission=True)
    def query_fleet(params: QueryFleetParams, invocation: ToolInvocation) -> ToolResult:
        items = store.list_workflows(
            phase=params.phase, agency=params.agency, has_exception=params.has_exception
        )
        excs = store.list_exceptions()
        by_phase: dict[str, int] = {}
        by_status: dict[str, int] = {}
        for w in items:
            by_phase[w.current_phase] = by_phase.get(w.current_phase, 0) + 1
            by_status[w.status] = by_status.get(w.status, 0) + 1
        result = {
            "total": len(items),
            "by_phase": by_phase,
            "by_status": by_status,
            "open_exception_count": len(excs),
            "recent_exceptions": [
                {"id": e.id, "workflow_id": e.workflow_id, "category": e.category, "severity": e.severity}
                for e in excs[-5:]
            ],
        }
        return ToolResult(text_result_for_llm=json.dumps(result), result_type="success")

    return query_fleet
